lua/wikis/ageofempires/Infobox/Upgrade/Custom.lua

```diff
@@ -0,0 +1,132 @@
+---
+-- @Liquipedia
+-- page=Module:Infobox/Upgrade/Custom
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Lua = require('Module:Lua')
+
+local Array = Lua.import('Module:Array')
+local Class = Lua.import('Module:Class')
+
+local Upgrade = Lua.import('Module:Infobox/Upgrade')
+local Injector = Lua.import('Module:Widget/Injector')
+
+local Widgets = Lua.import('Module:Widget/All')
+local Cell = Widgets.Cell
+local Center = Widgets.Center
+local Title = Widgets.Title
+local AgeIcon = Lua.import('Module:Widget/Infobox/AgeIcon')
+local ExpansionIcon = Lua.import('Module:Widget/Infobox/ExpansionIcon')
+local HtmlWidgets = Lua.import('Module:Widget/Html/All')
+local Link = Lua.import('Module:Widget/Basic/Link')
+
+local AGE_TITLE = {
+	['feudalage'] = 'Researched at age',
+	default = 'Starting age',
+}
+AGE_TITLE.castleage = AGE_TITLE.feudalage
+AGE_TITLE.imperialage = AGE_TITLE.feudalage
+
+---@class AoeUpgradeInfobox: UpgradeInfobox
+local CustomUpgrade = Class.new(Upgrade)
+
+---@class AoeUpgradeInfoboxWidgetInjector: WidgetInjector
+---@field caller AoeUpgradeInfobox
+local CustomInjector = Class.new(Injector)
+
+---@param frame Frame
+---@return Html
+function CustomUpgrade.run(frame)
+	local upgrade = CustomUpgrade(frame)
+	upgrade.args.informationType = 'Tech'
+	upgrade:setWidgetInjector(CustomInjector(upgrade))
+	return upgrade:createInfobox()
+end
+
+---@param id string
+---@param widgets Widget[]
+---@return Widget[]
+function CustomInjector:parse(id, widgets)
+	local args  = self.caller.args
+	if id == 'introduced' then
+		---@param introduced string?
+		---@return (string|Widget)[]?
+		local introducedDisplay = function(introduced)
+			if not introduced then return end
+			return {
+				ExpansionIcon{expansion = introduced},
+				HtmlWidgets.I{children = {Link{link = introduced}}},
+			}
+		end
+		return {
+			Cell{name = 'First introduced', children = introducedDisplay(args.introduced), options = {separator = ' '}},
+		}
+	elseif id == 'research' then
+		local civilizations = args.civilizations or args.civs
+
+		-- args.name usage is correct (copied from template code!)
```

Comment:
fwiw this is done because technically the ages are also technologies/upgrades you research. So this is to switch between the display for age upgrades and "normal" upgrades.

Could introduce a local `isAgeUpgrade` and use that to switch, might be a bit better to understand compared to the LUT.

Making it an extra parameter would also be fine for me, we can adjust the infoboxes if needed.

Reply:
decide on something you want and lmk
will implement it then^^